core/lib/types/src/contract_verification/contract_identifier.rs

```diff
@@ -0,0 +1,388 @@
+    /// Keccek256 hash of the metadata detected (only for EraVM).
+    Keccak256(H256),
+    /// CBOR metadata detected.
+    Cbor(H256),
+}
+
+impl DetectedMetadata {
+    pub fn hash(&self) -> H256 {
+        match self {
+            Self::Keccak256(hash) | Self::Cbor(hash) => *hash,
+        }
+    }
+}
+
+/// Possible values for the metadata hashes structure.
+/// Details can be found here: https://docs.soliditylang.org/en/latest/metadata.html
+///
+/// We're not really interested in the values here, we just want to make sure that we
+/// can deserialize the metadata.
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+struct CborMetadata {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    ipfs: Option<Vec<u8>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    bzzr1: Option<Vec<u8>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    bzzr0: Option<Vec<u8>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    experimental: Option<bool>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    solc: Option<Vec<u8>>,
+}
+
+impl ContractIdentifier {
+    pub fn from_bytecode(bytecode_marker: BytecodeMarker, bytecode: &[u8]) -> Self {
+        // Calculate the hash for bytecode with metadata.
+        let bytecode_sha3 = H256::from_slice(&keccak256(bytecode));
+        let mut self_ = Self {
+            bytecode_marker,
+            bytecode_sha3,
+            bytecode_without_metadata_sha3: None,
+            metadata_size: 0,
+        };
+
+        // For EraVM, the default metadata is keccak256 hash of the metadata.
+        // Try to use it as a default value (to be overridden if CBOR metadata is detected).
+        if bytecode_marker == BytecodeMarker::EraVm {
+            // For metadata, we might have padding: it takes either 32 or 64 bytes depending
+            // on whether the amount of words in the contract is odd, so we need to check
+            // if there is padding.
+            if bytecode.len() > 64 {
+                let bytecode_without_metadata =
+                    if bytecode[bytecode.len() - 64..bytecode.len() - 32] == [0u8; 32] {
+                        // Padding is present, strip it.
+                        self_.metadata_size = 64;
+                        &bytecode[..bytecode.len() - 64]
+                    } else {
+                        // No padding, strip metadata only.
+                        self_.metadata_size = 32;
+                        &bytecode[..bytecode.len() - 32]
+                    };
+                let hash = H256::from_slice(&keccak256(bytecode_without_metadata));
+                // This could be overridden if CBOR metadata is detected.
+                self_.bytecode_without_metadata_sha3 = Some(DetectedMetadata::Keccak256(hash));
+            }
+        }
+
+        // Try to detect CBOR metadata.
+
+        // Last two bytes is the length of the metadata in big endian.
+        if bytecode.len() < 2 {
+            return self_;
```

Comment:
Nit: I'd refactor this part into a separate method returning a `Result` with a simple error type (maybe `()` or `&'static str`), so that shortcuts during extraction are more idiomatic.

Reply:
I started refactoring it, and took a slightly different approach (that also doesn't require mutating the structure). Hope it's better now.